ykllvmwrap/src/jitmodbuilder.cc

```diff
@@ -665,66 +680,41 @@ class JITModBuilder {
 
     // Store the active frames vector and its length in a separate struct to
     // save arguments.
-    AllocaInst *ActiveFramesStruct = storeVecWithLength(
-        FailBuilder, ActiveFrameVec,
-        ConstantInt::get(PointerSizedIntTy, ActiveFrames.size()));
+    AllocaInst *ActiveFramesStruct = createAndFillStruct(
+        FailBuilder, {ActiveFrameVec, ConstantInt::get(PointerSizedIntTy,
+                                                       ActiveFrames.size())});
+
+    // Make more space to store the locations of the corresponding live AOT
+    // values for this guard failure.
+    size_t Offset = LiveAOTSize;
+    LiveAOTSize += sizeof(AOTInfo) * LiveValues.size();
+    LiveAOTAddr = (AOTInfo *)realloc(LiveAOTAddr, LiveAOTSize);
+    // Get a pointer to this guard failure's region in the memory block.
+    AOTInfo *CurrentRegion = (AOTInfo *)((char *)LiveAOTAddr + Offset);
```

Comment:
We should probably use C++ casts throughout.

Reply:
Can `LiveAOTAddr` be typed `AOTInfo *` from the outset to avoid the repeated casts?

Reply:
Also, since the size of `struct AOTInfo` is fixed, would it be better to keep track of not an offset, but a "next free element" index. Then you can do:

```
AOTInfo *CurrentRegion = LiveAOTAddr[Idx];
```

thus killing the cast to `char *`.

This would also go hand-in-hand with the `reallocarray` change I proposed.

Reply:
I could,  but then I would have to keep around another variable "Idx", because I still need "Offset" to pass it into the deopt call.

Reply:
Also, I find thinking about a "next free element" index more difficult than thinking about an offset. But that might just be me.

Reply:
OK, let's leave it for now.

Reply:
My comment about casts still stands.

Reply:
> Can `LiveAOTAddr` be typed `AOTInfo *` from the outset to avoid the repeated casts?

It is typed `AOTInfo *` from the outset. But `realloc` returns a `void *` so I need to cast it back to `AOTInfo *`.

Reply:
Sure, but why not use a C++ cast instead of a C cast? We are writing C++ code after all :)

Reply:
Fixed in [9f64524](https://github.com/ykjit/yk/pull/565/commits/9f64524a8c6ae37403c0a32a379e26fa9abd2270) and [3bab316](https://github.com/ykjit/yk/pull/565/commits/3bab31655c95d2572784bae2201be15289a2404d).

To be fair, this whole realloc business wouldn't have been necessary had I stuck to using a simple heap allocated vector. But that would have made @ltratt unhappy. I'm doing such splits here to please the both of you, my balls are touching the floor. :coconut: :laughing: